### PR TITLE
Modernize Running Process APIs 

### LINKED ISSUE #2690: Fix deprecated methods to getting timeline data
### 💻 Environment

Platform: macOS

### 📒 Description
Inside `get_focused_window_mac.cc`

Fix warnings about deprecated methods:
`'GetFrontProcess' is deprecated: first deprecated in macOS 10.9`
`'GetProcessForPID' is deprecated: first deprecated in macOS 10.9`
`'SameProcess' is deprecated: first deprecated in macOS 10.9`
`'CopyProcessName' is deprecated: first deprecated in macOS 10.9`

Appeared after #2689

### PR DESCRIPTION
**Pull request by NghiaTranUIT**
### 📒 Description
This PR modernize some deprecated APIs to fetch the running Process, Title and Running process for a Timeline Recording feature.

### 🕶️ Types of changes
**Breaking change** (fix or feature that would cause existing functionality to change)

### 🤯 List of changes
- [x] Introduce MacOSApplicationInfo.mm (obj-c++) for calling some objc code
- [x] Replace deprecated APIs with the new one

### 👫 Relationships
Closes #2690

#### Able to get the Title and Filename if ScreenRecording is granted (Catalina)
1. Enable Timeline Recording in Preference
2. Grant Screen Recording Permission in Security
3. Play around and check whether it's able to get the Title name + FileName of other apps

Screen Shot 2019-12-23 at 15 55 24

#### Able to get the Title if ScreenRecording is granted (Catalina)
1. Enable Timeline Recording in Preference
2. Don't grant Screen Recording Permission in Security
3. Play around and check whether it's able to get the Title name of other apps.
4. Make sure Filename is the same with Title Name
Screen Shot 2019-12-23 at 15 57 52

#### Work well in 10.14
1. Play around the app in 10.14 or older OS X
2. Make sure the app is able to get the Title + Filename as usual (No need grant permission)

### CODE REVIEW

File: src/lib/osx/Kopsik/MacOSApplicationInfo.mm [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+@interface MacOSApplicationUtility : NSObject
+
++ (NSRunningApplication *) getFrontApplication;
+@end
+
+@implementation MacOSApplicationUtility
+
++ (NSRunningApplication *) getFrontApplication
+{
+    NSArray<NSRunningApplication *> *runningApps = [[NSWorkspace sharedWorkspace] runningApplications];
```
Comment (NghiaTranUIT) [inline src/lib/osx/Kopsik/MacOSApplicationInfo.mm]: Replace 
```
extern OSErr GetFrontProcess(ProcessSerialNumber * PSN) 
```
with 
```
Use +[NSWorkspace runningApplications] and look for the entry with isActive == YES.
```
Ref: https://github.com/phracker/MacOSX-SDKs/blob/ef9fe35d5691b6dd383c8c46d867a499817a01b6/MacOSX10.15.sdk/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/HIServices.framework/Versions/A/Headers/Processes.h#L418-L434

File: src/lib/osx/Kopsik/MacOSApplicationInfo.mm [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        CFNumberRef window_owner_pid = (CFNumberRef)CFDictionaryGetValue(
+            dict, kCGWindowOwnerPID);
+        int window_owner_pid_int;
+        CFNumberGetValue(window_owner_pid, kCFNumberIntType,
+                         &window_owner_pid_int);
+
+        // Get app with pid
+        NSRunningApplication *app = [NSRunningApplication runningApplicationWithProcessIdentifier:window_owner_pid_int];
```
Comment (NghiaTranUIT) [inline src/lib/osx/Kopsik/MacOSApplicationInfo.mm]: Replace
```
GetProcessForPID()
```
with 
```
Use +[NSRunningApplication runningApplicationWithProcessIdentifier:].
```

https://github.com/phracker/MacOSX-SDKs/blob/ef9fe35d5691b6dd383c8c46d867a499817a01b6/MacOSX10.15.sdk/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/HIServices.framework/Versions/A/Headers/Processes.h#L901-L903

File: src/lib/osx/Kopsik/MacOSApplicationInfo.mm [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                if (!title->empty()) {
+                    break;
+                }
+            }
+        }
+    }
+    CFRelease(windows);
+
+    // get application filename
+    NSString *appName = frontApp.localizedName;
```
Comment (NghiaTranUIT) [inline src/lib/osx/Kopsik/MacOSApplicationInfo.mm]: Replace
```
CopyProcessName(&front_process_serial_number, &processName);
```
with
```
Use the localizedName property of the appropriate NSRunningApplication object.
```
https://github.com/phracker/MacOSX-SDKs/blob/ef9fe35d5691b6dd383c8c46d867a499817a01b6/MacOSX10.15.sdk/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/HIServices.framework/Versions/A/Headers/Processes.h#L824

**Comment by skel35** [comment]
@NghiaTranUIT Please rebase this PR to resolve conflicts correctly 🙇 
Now that I have 10.14 VM installed I can finally test this PR 💻

**Comment by NghiaTranUIT** [comment]
Rebased 🙌

**Pull request merged**